distributed/tests/test_scheduler.py

```diff
@@ -3538,3 +3541,40 @@ async def test_repr(s, a):
             repr(ws_b)
             == f"<WorkerState {b.address!r}, status: running, memory: 0, processing: 0>"
         )
+
+
+@gen_cluster(client=True)
+async def test_stimuli(c, s, a, b):
+    f = c.submit(inc, 1)
+    key = f.key
+
+    await f
+    await c.close()
+
+    assert_story(
+        s.story(key),
+        [
+            (key, "released", "waiting", {key: "processing"}),
+            (key, "waiting", "processing", {}),
+            (key, "processing", "memory", {}),
+            (
+                key,
+                "memory",
+                "forgotten",
+                {},
+            ),
+        ],
+    )
+
+    stimuli = [
+        "update-graph-hlg",
+        "update-graph-hlg",
+        "task-finished",
+        "client-releases-keys",
+    ]
```

Comment:
I'm curious, why are there two update-graph-hlg stimuli here?

Reply:
They're actually the same stimulus_id. Stimuli are of the form `f{name}-{time()}` so only the prefix is tested here. The stimulus_id is duplicated because there are 4 events in the scheduler transition log story.

```python
        [  # Expected subset of transition log
            (key, "released", "waiting", {key: "processing"}),
            (key, "waiting", "processing", {}),
            (key, "processing", "memory", {}),
            (key, "memory", "forgotten", {} ),
        ],
```

In the above case, a single `update-graph-{time()` stimulus is associated with (1) the transition from "released" to "waiting" (2) "waiting" to "processing". Once the task successfully completes on a worker, a new `task-finished-{time()}` stimulus_id is generated and sent to the scheduler, where (3) the task is transitioned from "processing" to "memory". Once the client has received the result, the scheduler releases the keys ("client-release-keys-time()") and (4) transitions the task from "memory" to "forgotten"